tracer/src/Datadog.Trace/DuckTyping/README.md

```diff
@@ -340,11 +340,106 @@ public interface IProxyMyHandlerConfiguration
 
 In this example the non public instance of `MyHandlerConfiguration` when calling the `Configuration` property is going to be wrapped with a `IProxyMyHandlerConfiguration` instance automatically. That allow us to access the internal data of that non public type.
 
+## Best practices
+
+Where possible, duck-type proxies should be implemented as an interface, e.g. `ISomeObject` _as long as they can be used in a constraint_. In these cases, the value passed to the method will be a struct which implements both `ISomeObject` and `IDuckType`. 
+
+> Note that the value passed to the integration will _never_ be `null` as it's a `struct`. If you need to check the original proxy value, you must add the `IDuckType` constraint also, and check `IDuckType.Instance` for `null`.
+
+In some cases it will not be possible to use the _constraint_ approach. In these cases, you should use a `[DuckCopy]` `struct` where possible. If you need to call _methods_ on the proxy, then you can't use a `struct` proxy, and instead should use an `interface` proxy.
+
+In summary:
+```
+┌────────────────────────┐            ┌────────────────┐
+│ Can the proxy be added │   Yes      │Use an Interface│
+│ as a constraint to the ├───────────►|     proxy      │
+│      integration?      │            └────────────────┘
+└───────────┬────────────┘
+            │ No
+            │
+┌───────────▼────────────┐   Yes      ┌────────────────┐
+│  Do you need to call   ├───────────►|Use an Interface│
+│ a method on the proxy? │            |      proxy     |    
+└───────────┬────────────┘            └────────────────┘
```

Comment:
I thought at this point we would want to use a class with virtual method definitions?

Reply:
Yes that was the previous advice, virtual method are a little bit faster than an interface method call. But I ran a new benchmark using the actual ducktype implementation and with all the options we have, and the difference is something like 1 or 2 nanoseconds, so I really think it doesn't worth.

This PR includes the results of the benchmarks and as a conclusion we should really try to use the generic constraints (20ns diff).

Reply:
Sounds good 👍🏼